introduce PARALLEL_TESTS for builder_run_tests (all builders)

fmessmer opened the pull request:
referring to issue https://github.com/ros-industrial/industrial_ci/issues/444
fixes #444 

as the assumption for the failures during `builder_run_tests` is "memory issue", I'm trying to limit the concurrent jobs by setting `--cmake-args -j1` for the `builder_run_tests` phase

for this, I introduced a new setting `TEST_CMAKE_ARGS` (to be distinguished from `*CMAKE_ARGS` for the `builder_run_build` phase)

by setting `TEST_CMAKE_ARGS=-j1`, I expect the command to be extended to 
```
catkin build --catkin-make-args run_tests -- --no-status --cmake-args -j1 --
```
which works fine without `industrial_ci`, i.e. running it in my workspace

however - and this is the reason for the `[WIP]` tag - when using
```
rosrun industrial_ci run_ci ROS_REPO=ros VERBOSE_OUTPUT=false VERBOSE_TESTS=false NOT_TEST_DOWNSTREAM=true CMAKE_ARGS=-DCMAKE_BUILD_TYPE=Release TEST_CMAKE_ARGS=-j1 ADDITIONAL_DEBS='apt-utils dialog openssh-client python-catkin-tools python-pip' DEBUG_BASH=true ROS_DISTRO=melodic DOCKER_IMAGE=ros:melodic-perception TARGET_WORKSPACE=.travis.rosinstall
```
I get the following failure:
```
+ exec catkin build --catkin-make-args run_tests -- --no-status --cmake-args -j1 --
----------------------------------------------------
Profile:                     default
Extending:          [cached] /opt/ros/melodic
Workspace:                   /root/target_ws
----------------------------------------------------
Build Space:        [exists] /root/target_ws/build
Devel Space:        [exists] /root/target_ws/devel
Install Space:      [exists] /root/target_ws/install
Log Space:          [exists] /root/target_ws/logs
Source Space:       [exists] /root/target_ws/src
DESTDIR:            [unused] None
----------------------------------------------------
Devel Space Layout:          linked
Install Space Layout:        merged
----------------------------------------------------
Additional CMake Args:       -j1
Additional Make Args:        None
Additional catkin Make Args: run_tests
Internal Make Job Server:    True
Cache Job Environments:      False
----------------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
----------------------------------------------------
Workspace configuration appears valid.

NOTE: Forcing CMake to run for each package.
----------------------------------------------------
[build] Found '1' packages in 0.0 seconds.                                                                       
Starting  >>> mojin_supported_robots               
_________________________________________________________________________________________________________________
Errors     << mojin_supported_robots:cmake /root/target_ws/logs/mojin_supported_robots/build.cmake.001.log
CMake Error: The source directory "/root/target_ws/build/mojin_supported_robots/-j1" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
cd /root/target_ws/build/mojin_supported_robots; catkin build --get-env mojin_supported_robots | catkin env -si  /usr/bin/cmake /root/target_ws/src/mojin_supported_robots --no-warn-unused-cli -DCATKIN_DEVEL_PREFIX=/root/target_ws/devel/.private/mojin_supported_robots -DCMAKE_INSTALL_PREFIX=/root/target_ws/install -j1; cd -
.................................................................................................................
Failed     << mojin_supported_robots:cmake          [ Exited with code 1 ]
Failed    <<< mojin_supported_robots                [ 0.1 seconds ]
[build] Summary: 0 of 1 packages succeeded.
[build]   Ignored:   None.
[build]   Warnings:  None.
[build]   Abandoned: None.
[build]   Failed:    1 packages failed.
[build] Runtime: 0.1 seconds total.
```

highlighting the main error line:
```
CMake Error: The source directory "/root/target_ws/build/mojin_supported_robots/-j1" does not exist.
```


@ipa-mdl 
 - do you have an idea why this fails together with `industrial_ci/master` branch? am I not escaping correctly?
 - would you be willing to merge something like this in order to work around the memory issue reported in #444?
 - would I need to adjust something for the other builders (colcon, etc...)?

ipa-mdl commented:
> do you have an idea why this fails together with industrial_ci/master branch? am I not escaping correctly?

I did not have a closer look, but I guess you have added the arguments to the wrong part of the commnad line.

> would you be willing to merge something like this in order to work around the memory issue reported in #444?

Yes and no.
I will not merge a features that allows to pass tool-specific arguments. Instead, I tried to implement semantic option, e.g. `VERBOSE_OUTPUT=true` instead if `..._ARGS=-vi`.

For now I would prefer to limit all tests to only one job per default / hard-coded.

> would I need to adjust something for the other builders (colcon, etc...)?

Yes, if possible.

ipa-mdl commented on industrial_ci/src/workspace.sh:
```diff
@@ -268,6 +268,13 @@ function ici_test_workspace {
+    ici_parse_env_array  test_cmake_args TEST_CMAKE_ARGS
+    local -a args
+    if [ ${#test_cmake_args[@]} -gt 0 ]; then
+        args+=(--cmake-args "${test_cmake_args[@]}" --)
```
This adds `--` twice, because `builder_test_results` uses it as well.

ipa-mdl commented:
> I'm trying to limit the concurrent jobs by setting --cmake-args -j1 for the builder_run_tests phase

I am not sure if `-j1` is a CMake option, I think it is a make option.

fmessmer commented:
rebased to latest `master` update

---

> I am not sure if `-j1` is a CMake option, I think it is a make option.

yes, it's a `make-args` option...changed it and now it works...

---

> > would you be willing to merge something like this in order to work around the memory issue reported in #444?
> 
> Yes and no.
> I will not merge a features that allows to pass tool-specific arguments. Instead, I tried to implement semantic option, e.g. `VERBOSE_OUTPUT=true` instead of `..._ARGS=-vi`.
> 
> For now I would prefer to limit all tests to only one job per default / hard-coded.

so, `TEST_MAKE_ARGS` is not acceptable...
would a re-activation of `ROS_PARALLEL_JOBS`/`ROS_TEST_PARALLEL_JOBS` (as in `legacy`) be acceptable?
or do you really want to hard-code `--make-args -j1 --`

---

> > would I need to adjust something for the other builders (colcon, etc...)?
> 
> Yes, if possible.

let's discuss this once we found an acceptable approach for `catkin-tools`

ipa-mdl commented:
> would a re-activation of ROS_PARALLEL_JOBS/ROS_TEST_PARALLEL_JOBS (as in legacy) be acceptable?

No, because these do not specify numbers, but the actual flags.
A limit of 1 job is very simple to implement, as soon as we have more, it gets complicated.
E.g. catkin-tools has a limit for the number packages as well.

fmessmer commented:
Ok, so this feature is the key for solving my issue https://github.com/ros-industrial/industrial_ci/issues/444
But your last comment confuses me and I do not really understand how you want me to implement it now....
 - not like it is proposed in this PR :heavy_check_mark: 
 - not using `ROS_PARALLEL_JOBS/ROS_TEST_PARALLEL_JOBS` :heavy_check_mark: 
 - hardcode without option to disable, i.e. always `run_tests` with `-j1`
 - option to enable/disable, i.e. something like `NOT_PARALLEL_TESTS=true/false` which sets `--make-args -j1 --` if `true` to `ici_exec_in_workspace "$extend" "$ws" catkin build --catkin-make-args run_tests -- "${opts[@]}" --no-status`?

ipa-mdl commented:
> hardcode without option to disable, i.e. always run_tests with -j1

This would be my starting point.

> option to enable/disable, i.e. something like NOT_PARALLEL_TESTS=true/false

Yes, as an extension.

fmessmer commented:
ok, I implemented it as `NOT_PARALLEL_TESTS`...

ipa-mdl commented on industrial_ci/src/builders/catkin_tools.sh:
```diff
@@ -40,6 +40,9 @@ function builder_run_tests {
     if [ "$IMMEDIATE_TEST_OUTPUT" == true ]; then
         opts+=("-i")
     fi
+    if [ "$NOT_PARALLEL_TESTS" == true ]; then
```
Now that I see it, I would prefer:
```
if [ "$PARALLEL_TESTS" != true ]; then
```
It is shorter and easier to understand ^^

ipa-mdl commented on industrial_ci/src/builders/catkin_tools.sh:
```diff
@@ -40,6 +40,9 @@ function builder_run_tests {
         opts+=("-i")
     fi
+    if [ "$NOT_PARALLEL_TESTS" == true ]; then
+        opts+=("-j1")
```
The same flag should work for `catkin_make` and `catkin_make_isolated`

ipa-mdl replied:
For catkin-tools, adding `-p 1` might make sense as well.

ipa-mdl commented:
For `colcon` it should be `'--executor sequential` and perhaps `-j1`.

fmessmer commented:
I don't have proper tests for the other builders
Do you want to add a test for this option to `industrial_ci`'s `.travis.yml`? Any suggestions?

ipa-mdl commented:
Thanks for the update!

> Do you want to add a test for this option to industrial_ci's .travis.yml?

No, not for now.
I will test it with `run_ci`.
I don't to add more tests now without restructuring first.

ipa-mdl commented on doc/index.rst:
```diff
@@ -200,6 +200,7 @@ Note that some of these currently tied only to a single option, but we still lea
 * **NOT_TEST_DOWNSTREAM** (default: not set): If true, tests in the downstream workspace won't be run.
 * **OS_CODE_NAME** (default: derived from ROS_DISTRO): See `this section for the detail <https://github.com/ros-industrial/industrial_ci/blob/master/doc/index.rst#optional-type-of-os-and-distribution>`__.
 * **OS_NAME** (default: ubuntu): Possible options: {`ubuntu`, `debian`}. See `this section for the detail <https://github.com/ros-industrial/industrial_ci/blob/master/doc/index.rst#optional-type-of-os-and-distribution>`__.
+* **PARALLEL_TESTS** (default: not set): If `false`, build tool (e.g. Catkin) executes `run_tests` step sequentially, i.e. only one job at a time (equal to `-j1`).
```
```suggestion
* **PARALLEL_TESTS** (default: not set): If `true`, build tool (e.g. Catkin) executes `run_tests` jobs in parallel, instead of only one job at a time.
```

ipa-mdl commented on industrial_ci/src/builders/catkin_make_isolated.sh:
```diff
@@ -31,7 +31,11 @@ function builder_run_build {
+    if [ "$PARALLEL_TESTS" == false ]; then
+        opts+=("-j1")
+    fi
+    _run_catkin_make_isolated run_tests "$1" "$2" "$opts"
```
```suggestion
    _run_catkin_make_isolated run_tests "$1" "$2" "${opts[@]}"
```

fmessmer commented:
it's been late yesterday....but I guess, eventually, you are going to "squash-merge" anyway :wink:

ipa-mdl commented on industrial_ci/src/builders/colcon.sh:
```diff
@@ -40,7 +40,13 @@ function builder_run_tests {
-    ici_exec_in_workspace "$extend" "$ws" colcon test --event-handlers "${_colcon_event_handlers[@]}" "${output_handler}"
+    local -a opts
+    if [ "$PARALLEL_TESTS" == false ]; then
+        opts+=("--executor sequential --parallel-workers 1")
```
```suggestion
        opts+=("--executor sequential)
```
Nitpick.

ipa-mdl commented:
Code LGTM. Let's see if Travis agrees.
Afterwards I will run my local tests again and then it should ready for squash-merge.

---
```bash
for p in false true; do for b in catkin_tools catkin_make catkin_make_isolated colcon; do echo BUILDER=$b PARALLEL_TESTS=$p; time ~/build/industrial_ci/industrial_ci/scripts/run_ci ROS_DISTRO=melodic BUILDER=$b PARALLEL_TESTS=$p; done; done
```

ipa-mdl commented on industrial_ci/src/builders/colcon.sh:
```diff
@@ -40,7 +40,13 @@ function builder_run_tests {
+    else
+        opts+=("--executor parallel")
+    fi
+    ici_exec_in_workspace "$extend" "$ws" colcon test --event-handlers "${_colcon_event_handlers[@]}" "${output_handler}" "${opts}"
```
```suggestion
    ici_exec_in_workspace "$extend" "$ws" colcon test  "${opts[@}"
```
"${var[@}" expands the array into individual strings.

ipa-mdl commented on industrial_ci/src/builders/colcon.sh:
```diff
@@ -40,7 +40,13 @@ function builder_run_tests {
         output_handler="console_cohesion+"
     fi
-    ici_exec_in_workspace "$extend" "$ws" colcon test --event-handlers "${_colcon_event_handlers[@]}" "${output_handler}"
+    local -a opts
```
```suggestion
    local opts=(--event-handlers "${_colcon_event_handlers[@]}" "${output_handler}")
```

ipa-mdl commented on industrial_ci/src/builders/colcon.sh:
```diff
@@ -40,7 +40,13 @@ function builder_run_tests {
-    ici_exec_in_workspace "$extend" "$ws" colcon test --event-handlers "${_colcon_event_handlers[@]}" "${output_handler}"
+    local opts=(--event-handlers "${_colcon_event_handlers[@]}" "${output_handler}")
+    if [ "$PARALLEL_TESTS" == false ]; then
+        opts+=("--executor sequential")
```
```suggestion
        opts+=(--executor sequential)
```

ipa-mdl commented on industrial_ci/src/builders/colcon.sh:
```diff
@@ -40,7 +40,13 @@ function builder_run_tests {
+    if [ "$PARALLEL_TESTS" == false ]; then
+        opts+=("--executor sequential")
+    else
+        opts+=("--executor parallel")
```
```suggestion
        opts+=(--executor parallel)
```

fmessmer replied:
why do we need to remove the quotes?
should this be done for all occurrences, i.e.
 - https://github.com/ros-industrial/industrial_ci/pull/446/files#diff-46bbdbaf63f42dea8f37320cc6574e1aR33
 - https://github.com/ros-industrial/industrial_ci/pull/446/files#diff-9fbeddd0a763e900f5b2109a24aceb2aR36
 - https://github.com/ros-industrial/industrial_ci/pull/446/files#diff-807dad9df09c471416c53ddfe3af7795R38
 - https://github.com/ros-industrial/industrial_ci/pull/446/files#diff-807dad9df09c471416c53ddfe3af7795R41
 - https://github.com/ros-industrial/industrial_ci/pull/446/files#diff-807dad9df09c471416c53ddfe3af7795R44
 - https://github.com/ros-industrial/industrial_ci/pull/446/files#diff-60272b6836288cfae2c2fabe8971649fR45
 - https://github.com/ros-industrial/industrial_ci/pull/446/files#diff-60272b6836288cfae2c2fabe8971649fR47

ipa-mdl replied:
> why do we need to remove the quotes?

because with quote it one string `"--executor parallel"`, without it is `"--executor"``and `"parallel"`, which is needed for proper parsing.

> should this be done for all occurrences, i.e.

Single strings are fine. I will have another look.

ipa-mdl commented on industrial_ci/src/builders/catkin_tools.sh:
```diff
@@ -40,6 +40,9 @@ function builder_run_tests {
         opts+=("-i")
     fi
+    if [ "$PARALLEL_TESTS" == false ]; then
+        opts+=("-j1 -p1")
```
```suggestion
        opts+=(-j1 -p1)
```

ipa-mdl commented on industrial_ci/src/builders/catkin_make.sh:
```diff
@@ -28,7 +28,11 @@ function builder_run_build {
-    ici_exec_in_workspace "$extend" "$ws" catkin_make --make-args run_tests "$@"
+    local -a opts
+    if [ "$PARALLEL_TESTS" == false ]; then
+        opts+=("-j1")
```
```suggestion
        opts+=(-j1)
```
Not needed, but to make it alike.

ipa-mdl commented on industrial_ci/src/builders/catkin_make_isolated.sh:
```diff
@@ -31,7 +31,11 @@ function builder_run_build {
-    _run_catkin_make_isolated run_tests "$1" "$2"
+    local -a opts
+    if [ "$PARALLEL_TESTS" == false ]; then
+        opts+=("-j1")
```
```suggestion
        opts+=(-j1)
```
